pygfx/utils/transform.py

```diff
@@ -0,0 +1,244 @@
+import numpy as np
+import pylinalg as pla
+from datetime import datetime
+
+from typing import List
+
+
+class cached:
+    """Cache for computed properties.
+
+    This descriptor implements a minimal timestamp-based cache for computed
+    properties. The value of the property is computed using ``update_fn`` and
+    the result is cached until ``obj.last_modified`` advances. At this point the
+    value of the computed property is recomputed upon the next read.
+
+
+
+    Example
+    -------
+
+    class Foobar:
+        def __init__(self, x):
+            self._value = 0
+            self.last_modified = datetime.now()
+
+        @property
+        def value(self):
+            return self._value
+
+        @value.setter
+        def value(self, new_value):
+            self._value = new_value
+            self.last_modified = datetime.now()
+
+        @property
+        @cached
+        def double(self):
+            print("Computed value of double..")
+            return 2*x
+
+    """
+
+    def __init__(self, update_fn=None) -> None:
+        self.update_fn = update_fn
+        self.last_updated = None
+        self.cached_value = None
+
+    def __get__(self, obj: "AffineBase", objtype=None):
+        if obj is None:
+            return self
+
+        if self.last_updated is None or obj.last_modified > self.last_updated:
+            self.cached_value = self.update_fn(obj)
+            self.last_updated = obj.last_modified
+
+        return self.cached_value
+
+    def __call__(self, obj):
+        """Helper to allow interoperability with @property."""
+        return self.__get__(obj)
+
+
+class AffineBase:
+    last_modified: datetime
+    matrix: np.ndarray
+
+    @property
+    @cached
+    def inverse_matrix(self):
+        return np.linalg.inv(self.matrix)
+
+    @property
+    @cached
+    def position(self) -> np.ndarray:
+        pos, _, _ = pla.matrix_decompose(self.matrix)
+        return pos
+
+    @property
+    @cached
+    def rotation(self) -> np.ndarray:
+        _, rot, _ = pla.matrix_decompose(self.matrix)
+        return rot
+
+    def __array__(self, dtype=None):
+        return self.matrix.astype(dtype)
+
+    @property
+    @cached
+    def scale(self) -> np.ndarray:
+        _, _, scale = pla.matrix_decompose(self.matrix)
+        return scale
+
+
+class AffineTransform(AffineBase):
```

Comment:
I think this class is missing a setter for the matrix that then updates position, rotation scale

Reply:
> I think this class is missing a setter for the matrix that then updates position, rotation scale

We don't need to update position, rotation, and scale when the matrix changes. Only `last_modified`, i.e. flag that matrix has moved on / changed. Position, rotation, and scale are be updated by their getters if necessary.

The problem with adding a setter for `matrix` is that it only monitors changes to the full matrix/buffer, not changes of elements inside the matrix or in-place updates. This is why I added `AffineMatrix.flag_update` to allow consumers of `AffineMatrix` to communicate that they changed the matrix in place. I'm not 100% happy with the solution, but the only alternative I can think of is to subclass `np.ndarray` and add a monitor on `__setitem__`. We could do this, but I don't know how happy you would be with the complexity of this approach.

Reply:
> Position, rotation, and scale are be updated by their getters if necessary.

Ah, I didn't realize that. It's in the base class. Duh!

Reply:
> The problem with adding a setter for `matrix` is that it only monitors changes to the full matrix/buffer, not changes of elements inside the matrix or in-place updates. This is why I added `AffineMatrix.flag_update` to allow consumers of `AffineMatrix` to communicate that they changed the matrix in place. I'm not 100% happy with the solution, but the only alternative I can think of is to subclass `np.ndarray` and add a monitor on `__setitem__`. We could do this, but I don't know how happy you would be with the complexity of this approach.

Well, there's going to be a whole class of users that will exclusively write to the matrix property to transform an object. In three.js there's [a flag](https://threejs.org/docs/index.html#api/en/core/Object3D.matrixAutoUpdate) that you can set to "inform" object 3d that you don't want the matrix to be updated based on position/scale/rotation, but that you will be setting it yourself.

Does that maybe inspire some new ideas?

Reply:
Best I can think of right now is to make `AffineBase.matrix` a `@property` that returns a read-only view and to add a setter that updates the values of the underlying buffer in-place. This keeps the buffer itself intact (which I would prefer because of some world-matrix caching ideas that I still need to implement), allows updating the full matrix, and prevents in-place modifications that we can't detect.

The problem is precisely that it prevents in-place modification, which means that some operations on the matrix are not possible even though a user might wish to do them, e.g. passing `WorldObject.transform.matrix` to the `out` argument of pylinalg...

Reply:
I'll let you think on that one...

Reply:
Sounds like a viable approach to prevent users to shoot themselves in the foot 👍